File: Algorithm_2.2/22series-labs/Lab 6 - Bucket, 3-Way-Merge, Matrix Multiplication/bucket_sort.py
def insertion(A):
    n = len(A)
    for i in range(1,n):
        temp = A[i]
        j = i - 1
        while j >= 0 and temp < A[j]:
            A[j + 1] = A[j]
            j -= 1
        A[j + 1] = temp

def bucket(A):
    n = len(A)
    B = [[] for _ in range(10)]
    for i in range(n):
        B[int(A[i]*n)].append(A[i])
    for i in range(n):
        insertion(B[i])
    C = []
    for i in range(n):
        for x in B[i]:
            C.append(x)
    print(C)

A = [0.78,0.72,0.68,0.52,0.55,0.4,0.91,0.32,0.11,0.09]
bucket(A)

File: Algorithm_2.2/mySortingsCode/bucketSort.cpp
#include <bits/stdc++.h>
using namespace std;

// void InsertionSort(vector<float>& v) {
//     //to use: InsertionSort(bucket[i]);
//     for(int i = 1; i < v.size(); i++) {
//         float x = v[i];
//         int j = i - 1;

//         while(j >= 0 && v[j] > x) {
//             v[j + 1] = v[j];
//             j--;
//         }

//         v[j + 1] = x;
//     }
// }



vector<float> BucketSort(vector<float>& vec) {
    int n = vec.size() - 1;

    // 10 buckets: B0 ... B9
    vector<vector<float>> bucket(10);

    // Distribution
    for(int i = 1; i <= n; i++) {
        int idx = (int)(vec[i] * 10);      // floor(value * 10)

        // Safety (if value == 1.0)
        if(idx == 10) idx = 9;

        bucket[idx].push_back(vec[i]);
    }

    // Sort each bucket
    for(int i = 0; i < 10; i++) {
        sort(bucket[i].begin(), bucket[i].end());
        // Abdul Bari uses Insertion Sort.
        // STL sort is used here for simplicity.
    }

    // Concatenate
    vector<float> ans(n + 1);
    int k = 1;

    for(int i = 0; i < 10; i++) {
        for(float x : bucket[i]) {
            ans[k++] = x;
        }
    }

    return ans;
}

int main() {
    int n;
    cin >> n;

    vector<float> vec(n + 1);

    for(int i = 1; i <= n; i++) {
        cin >> vec[i];
    }

    vec = BucketSort(vec);

    for(int i = 1; i <= n; i++) {
        cout << fixed << setprecision(2) << vec[i] << " ";
    }

    cout << endl;

    return 0;
}